src/_guides/language/effective-dart/design.md

```diff
@@ -1005,90 +1089,254 @@ var names = people.map((Person person) {
+`dynamic` when it is implicit. With the new stronger type system and type
+inference, users now expect Dart to behave like an inferred statically-typed
+language. With that mental model, it is an unpleasant surprise to discover that
+a region of code has silently lost all of the safety and performance of static
+types.
+
+</aside>
+
+
+### PREFER signatures in function type annotations.
+
+The identifier `Function` by itself without any return type or parameter
+signature refers to the special [Function][] type. This type is only
+marginally more useful than using `dynamic`. If you're going to annotate, prefer
+a full function type that includes the parameters and return type of the
+function.
+
+[Function]: {{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-core/Function-class.html
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-Function)" replace="/(void )?Function(\(.*?\))?/[!$&!]/g"?>
+{% prettify dart %}
+bool isValid(String value, bool [!Function(String string)!] test) => ...
 {% endprettify %}
 
 {:.bad-style}
-<?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-Function)" replace="/Function/[!$&!]/g"?>
+{% prettify dart %}
+bool isValid(String value, [!Function!] test) => ...
+{% endprettify %}
+
+[fn syntax]: #prefer-inline-function-types-over-typedefs
+
+One exception to this guideline is if you want a type that represents the union
+of multiple different function types. For example, you may accept a function
+that takes one parameter or a function that takes two. Since we don't have union
+types, there's no way to precisely type that and you'd normally have to use
+`dynamic`. `Function` is at least a little more helpful than that:
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (function-arity)" replace="/(void )?Function(\(.*?\))?/[!$&!]/g"?>
 {% prettify dart %}
-dynamic lookUpOrDefault(String name, Map map, dynamic defaultValue) {
-  var value = map[name];
-  if (value != null) return value;
-  return defaultValue;
+void handleError([!void Function()!] operation, [!Function!] errorHandler) {
+  try {
+    operation();
+  } catch (err, stack) {
+    if (errorHandler is [!Function(Object)!]) {
+      errorHandler(err);
+    } else if (errorHandler is [!Function(Object, Object)!]) {
+      errorHandler(err, stack);
+    }
+  }
 }
 {% endprettify %}
 
 
-### AVOID annotating with `Function`.
+### DON'T specify a return type for a setter.
 
-The `Function` type is barely more precise than using no annotation at all. If
-you're bothering to annotate, it's better to use a precise function type that
-describes the signature and return type of the function.
+Setters always return `void` in Dart. Writing the word is pointless.
 
-If you are annotating a field, this does mean you have to create a typedef, but
-that's usually worth doing.
+{:.bad-style}
```

Comment:
Why isn't the good example before the bad example?

Reply:
I think the context for this comment may have been changed.